src/AggregateFunctions/AggregateFunctionLast2Samples.h

```diff
@@ -0,0 +1,387 @@
+#pragma once
+
+#include <cstddef>
+#include <cstring>
+#include <memory>
+#include <type_traits>
+
+#include <IO/WriteHelpers.h>
+#include <IO/ReadHelpers.h>
+
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/DataTypesDecimal.h>
+#include <DataTypes/DataTypeTuple.h>
+#include <Columns/ColumnVector.h>
+#include <Columns/ColumnTuple.h>
+
+#include <AggregateFunctions/IAggregateFunction.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int INCORRECT_DATA;
+}
+
+struct Settings;
+
+/// Aggregate function to store last (latest by timestamp) 2 samples with distinct timestamps
+template <typename TimestampType, typename ValueType>
+class AggregateFunctionLast2Samples final :
+    public IAggregateFunctionHelper<AggregateFunctionLast2Samples<TimestampType, ValueType>>
+{
+public:
+    static constexpr bool DateTime64Supported = true;
+
+    using Base = IAggregateFunctionHelper<AggregateFunctionLast2Samples<TimestampType, ValueType>>;
+
+    using ColVecType = ColumnVectorOrDecimal<TimestampType>;
+    using ColVecResultType = ColumnVectorOrDecimal<ValueType>;
+
+    String getName() const override
+    {
+        return "timeSeriesLastTwoSamples";
+    }
+
+    /// Stores two samples with most recent distinct timestamps
+    /// If there are two samples with the same timestamp, the one with bigger value is stored
+    struct Data
+    {
+        ValueType values[2];            /// In common scenario values are Float64, so put them first as they need 8-byte alignment
+        TimestampType timestamps[2];    /// Timestamps might be stored as DateTime64, DateTime32 or even as 16-bit delta from the base timestamp of the grid
+        UInt16 filled = 0;              /// Number of samples stored: 0, 1 or 2
+
+        void add(TimestampType timestamp, ValueType value)
+        {
+            if (filled == 0)
+            {
+                timestamps[0] = timestamp;
+                values[0] = value;
+                filled = 1;
+            }
+            else if (filled == 1)
+            {
+                if (timestamp > timestamps[0])
+                {
+                    /// Move the first sample to the second one
+                    timestamps[1] = timestamps[0];
+                    values[1] = values[0];
+                    /// Set the first sample to the new one
+                    timestamps[0] = timestamp;
+                    values[0] = value;
+
+                    filled = 2;
+                }
+                else if (timestamp == timestamps[0])
+                {
+                    /// Replace the value with bigger one
```

Comment:
Why do we replace the value with bigger one?

Reply:
This is done for stability. Normally, there should not be samples with equal timestamps but if for some reason they are present in user data we want to always choose the same one no mater in which order the data is scanned.